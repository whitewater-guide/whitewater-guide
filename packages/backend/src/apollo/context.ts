import * as koa from 'koa';
import { get } from 'lodash';
import { LastMeasurementLoader } from '../features/measurements/data-loader';
import { PurchasesLoader } from '../features/purchases/data-loader';
import { LANGUAGES } from '../ww-commons';

export interface ContextUser {
  id: string;
  admin: boolean;
  [key: string]: any;
}

export interface Context {
  language: string;
  user?: ContextUser;
  lastMeasurementLoader: LastMeasurementLoader;
  purchasesLoader: PurchasesLoader;
}

export const newContext = (ctx: Partial<koa.Context>): Context => {
  const user: ContextUser | undefined = ctx.state && ctx.state.user;
  const language = ctx.headers['x-editor-language'] ||
    get(user, 'language') ||
    ctx.acceptsLanguages!(LANGUAGES) ||
    'en';
  // Side-effect. Set response content-language
  ctx.set!('Content-Language', language);
  return {
    user,
    language,
    lastMeasurementLoader: new LastMeasurementLoader(),
    purchasesLoader: new PurchasesLoader(user, language),
  };
};
