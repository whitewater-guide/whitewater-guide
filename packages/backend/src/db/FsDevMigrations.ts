// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { FsMigrations } from 'knex/lib/migrate/sources/fs-migrations';

// When restoring production dump on development, migration names will have '.js' extension
// Therefore, development migrations with '.ts' will conflict them
// this is the workaround
// https://github.com/knex/knex/blob/master/lib/migrate/sources/fs-migrations.js
export class FsDevMigrations extends FsMigrations {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(
    migrationDirectories: string | string[],
    sortDirsSeparately?: boolean,
    loadExtensions?: string[],
  ) {
    super(migrationDirectories, sortDirsSeparately, loadExtensions);
  }

  getMigrationName(migration: any): string {
    return migration.file.replace('.ts', '.js');
  }
}
