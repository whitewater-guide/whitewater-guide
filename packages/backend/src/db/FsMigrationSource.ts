import fs from 'node:fs/promises';
import path from 'node:path';

import type { Knex } from 'knex';

// https://knexjs.org/guide/migrations.html#custom-migration-sources
export default class FsMigrationSource implements Knex.MigrationSource<string> {
  // eslint-disable-next-line node/no-process-env
  private readonly ext = process.env.NODE_ENV === 'production' ? 'js' : 'ts';
  private readonly directory: string;

  constructor(directory: string) {
    this.directory = directory;
  }
  // Must return a Promise containing a list of migrations.
  // Migrations can be whatever you want,
  // they will be passed as arguments to getMigrationName
  // and getMigration
  //
  // In our case migrations are filenames without extension
  async getMigrations(): Promise<string[]> {
    // In this example we are just returning migration names
    const dirContent = await fs.readdir(this.directory);
    // assume that migrations have extensions (== no directories with '.' symbol)
    return dirContent
      .filter((f) => f.includes('.'))
      .map((f) => f.split('.')[0])
      .sort();
  }

  // When restoring production dump on development, migration names will have '.js' extension
  getMigrationName(migration: string): string {
    return migration + '.js';
  }

  async getMigration(migration: string): Promise<Knex.Migration> {
    const moduleName = path.resolve(
      process.cwd(),
      this.directory,
      `${migration}.${this.ext}`,
    );
    return import(moduleName);
  }
}
