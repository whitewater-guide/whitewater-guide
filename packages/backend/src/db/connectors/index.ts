export * from './createConnectors';
export * from './OffsetConnector';
export * from './types';
