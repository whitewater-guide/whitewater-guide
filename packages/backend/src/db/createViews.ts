import { readFileSync } from 'fs';
import Knex from 'knex';
import path from 'path';

export async function createView(db: Knex, view: string, version = 1) {
  for (let i = version; i >= 1; i--) {
    const file = path.resolve(
      __dirname,
      '../migrations',
      `${i.toString(10).padStart(3, '0')}/${view}_view.sql`,
    );
    let sql: string;
    try {
      sql = readFileSync(file).toString();
    } catch (e) {
      continue;
    }
    await db.schema.raw(sql);
    return;
  }
  throw new Error(
    `No sql files for view '${view}' of version ${version} or lower found`,
  );
}

// Views will be created in reverse order
export async function createViews(
  db: Knex,
  version: number,
  ...views: string[]
) {
  const inverted = [...views].reverse();
  for (const view of inverted) {
    await createView(db, view, version);
  }
}
