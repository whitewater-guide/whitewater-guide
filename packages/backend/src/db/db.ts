/* eslint-disable node/no-process-env */
import Knex from 'knex';

import knexConfig from './knexfile';

const env = process.env.NODE_ENV || 'development';
const config: Knex.Config = knexConfig[env];

export const knex = Knex(config);

let testTransaction: Knex.Transaction | null = null;

export function db(getKnexInstance = false): Knex<any, any> {
  if (env === 'test' && !getKnexInstance) {
    if (testTransaction) {
      return testTransaction;
    }
    throw new Error('No saved test transaction found');
  }
  return knex;
}

// See here https://github.com/tgriesser/knex/issues/2076
// And here https://gist.github.com/odigity/7f37077de74964051d45c4ca80ec3250
export const holdTransaction = () => {
  if (env !== 'test') {
    throw new Error('holdTransaction is only available in test mode');
  }
  const tmp: any = {};
  const p = new Promise((resolve) => {
    tmp.resolve = resolve;
  });
  knex
    .transaction((tx) => {
      testTransaction = tx;
      tmp.resolve();
    })
    .catch(() => {
      // ignore
    });
  return p;
};

export const rollbackTransaction = async () => {
  if (env !== 'test') {
    throw new Error('rollbackTransaction is only available in test mode');
  }
  if (testTransaction) {
    await testTransaction.rollback();
    testTransaction = null;
  }
};
