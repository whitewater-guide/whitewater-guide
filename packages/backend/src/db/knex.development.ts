import knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
    keepAlive: true,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/development',
  },
  debug: process.env.DATABASE_DEBUG === 'true',
} as any;

module.exports = development;
