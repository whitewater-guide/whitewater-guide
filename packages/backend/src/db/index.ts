import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './runSqlFile';
export * from './createViews';
export * from './dropViews';
export * from './rawUpsert';
export * from './types';
export * from './db';
export * from './queryBuilders';
