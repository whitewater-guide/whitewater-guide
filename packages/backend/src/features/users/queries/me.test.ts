import { holdTransaction, rollbackTransaction } from '../../../db';
import { adminContext, anonContext } from '../../../test/context';
import { runQuery } from '../../../test/db-helpers';
import { User } from '../../../ww-commons';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const query = `
{
  me {
    id
    name
    avatar
    role
    email
    createdAt
    updatedAt
    language
    imperial
    editorSettings {
      language
    }
  }
}
`;

test('should return null for anon', async () => {
  const result = await runQuery(query, undefined, anonContext);
  expect(result.errors).toBeUndefined();
  expect(result.data).toBeDefined();
  expect(result.data!.me).toBeNull();
});

test('should user', async () => {
  const result = await runQuery(query, undefined, adminContext);
  expect(result.errors).toBeUndefined();
  const me: User = result.data!.me;
  expect(me).toMatchSnapshot();
});
