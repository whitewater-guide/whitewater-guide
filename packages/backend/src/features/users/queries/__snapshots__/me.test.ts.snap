// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return facebook user 1`] = `
Object {
  "accounts": Array [
    Object {
      "id": "__fb_admin_id__",
      "provider": "facebook",
    },
    Object {
      "id": "__apple_admin_id__",
      "provider": "apple",
    },
  ],
  "admin": true,
  "avatar": "https://ww.guide/pic.jpg",
  "createdAt": 2017-02-01T00:00:00.000Z,
  "editor": false,
  "editorSettings": null,
  "email": "kaospostage@gmail.com",
  "id": "bed59990-749d-11e7-8cf7-a6006ad3dba0",
  "imperial": false,
  "language": "en",
  "name": "Ivan Ivanov",
  "updatedAt": 2017-02-01T00:00:00.000Z,
  "verified": true,
}
`;

exports[`should return local user 1`] = `
Object {
  "accounts": Array [],
  "admin": false,
  "avatar": null,
  "createdAt": 2017-05-04T00:00:00.000Z,
  "editor": false,
  "editorSettings": null,
  "email": "konstantin@gmail.com",
  "id": "fa3ce7ba-36ab-11e8-b467-0ed5f89f718b",
  "imperial": false,
  "language": "ru",
  "name": "Another usr",
  "updatedAt": 2017-05-04T00:00:00.000Z,
  "verified": true,
}
`;
